src/rrd/mod.rs

```diff
@@ -0,0 +1,493 @@
+        let mut connection = self.pool.acquire().await?;
+        let timestamps: Vec<i64> = sqlx::query_scalar("SELECT DISTINCT timestamp FROM usage;")
+            .fetch_all(&mut connection)
+            .await?;
+        for ts in timestamps {
+            RRDImpl::print_ts(self, ts, &mut writer).await?;
+        }
+        Ok(writer)
+    }
+
+    async fn print_last_usage<W: Write>(&mut self, mut writer: W) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        writer.write_fmt(format_args!(".last\n"))?;
+        let records: Vec<(String, f64)> = sqlx::query_as("SELECT metric, usage FROM last;")
+            .fetch_all(&mut connection)
+            .await?;
+        for (metric, usage) in records {
+            writer.write_fmt(format_args!("\t{}: {}\n", metric, usage))?
+        }
+        Ok(())
+    }
+
+    async fn print_ts<W: Write>(&mut self, ts: i64, mut writer: W) -> Result<()> {
+        let mut connection = self.pool.acquire().await?;
+        let records: Vec<(String, f64)> =
+            sqlx::query_as("SELECT metric, usage FROM usage WHERE timestamp = ? ;")
+                .bind(ts)
+                .fetch_all(&mut connection)
+                .await?;
+        for (metric, usage) in records {
+            writer.write_fmt(format_args!("\t{}: {}\n", metric, usage))?
+        }
+        Ok(())
+    }
+
+    async fn retain(&mut self, now: i64) -> Result<()> {
+        // should retain be unsigned?
+        let mut connection = self.pool.acquire().await?;
+        let retain = (now - self.retention) as i64;
+        sqlx::query("DELETE FROM usage WHERE timestamp <= ? ;")
+            .bind(retain)
+            .execute(&mut connection)
+            .await?;
+        Ok(())
+    }
+
+    pub async fn slots(&mut self) -> Result<Vec<i64>> {
+        let mut connection = self.pool.acquire().await?;
+        let timestamps: Vec<i64> = sqlx::query_scalar("SELECT DISTINCT timestamp FROM usage ;")
+            .fetch_all(&mut connection)
+            .await?;
+        Ok(timestamps)
+    }
+
+    pub async fn slot_at(&mut self, ts: i64) -> Result<RRDSlotImpl> {
+        let connection = self.pool.acquire().await?;
+        let ts = (ts / self.window) * self.window;
+        RRDImpl::retain(self, ts).await?;
+        Ok(RRDSlotImpl {
+            connection,
+            key: ts,
+        })
+    }
+}
+
+impl RRDSlotImpl {
+    async fn get_last(&mut self, key: &str) -> Result<Option<f64>> {
+        let last: Vec<f64> = sqlx::query_scalar("SELECT value FROM last WHERE metric = ? ;")
+            .bind(key)
+            .fetch_all(&mut self.connection)
+            .await?;
+        if last.is_empty() {
+            return Ok(None);
+        }
+        Ok(Some(last[0]))
+    }
+
+    async fn set_last(&mut self, key: &str, value: &f64) -> Result<()> {
+        sqlx::query("REPLACE INTO last (timestamp, metric, value) VALUES (?, ?, ?);")
+            .bind(self.key)
```

Comment:
The `timestamp` added to this `last` table is totally not related to the slot instead it's the time of the reporting. Hence this method is totally not related to the `Slot`.